### DISCUSSION TITLE
vim should have +clipboard by default

### DISCUSSION THREAD

**Discussion by kitmonisit** [Homebrew/discussions#1392] [Linux]
Hi, I think vim should have access to the X clipboard by default, i.e.

```
$ vim --version | grep clipboard
+clipboard
+xterm_clipboard
```

This will allow copying and pasting using `"*y` and `"*p` in Normal Mode from other applications.

I move to update the bottle such that vim is compiled this way. In my recent test (not a clean Linuxbrew install), the only additional build dependencies are `libx11` and `libxt`.

Note the `--with-x` option
```
    system "./configure", "--prefix=#{homebrew_prefix}",
                          "--mandir=#{man}",
                          "--enable-multibyte",
                          "--with-tlib=ncurses",
                          "--with-compiledby=homebrew",
                          "--enable-cscope",
                          "--enable-terminal",
                          "--enable-perlinterp",
                          "--enable-rubyinterp",
                          "--enable-python3interp",
                          "--enable-gui=no",
                          "--with-x",
                          "--enable-luainterp",
                          "--with-lua-prefix=#{formula["lua"].opt_prefix}"
```

**Comment by danielnachun** [comment]
We've discussed this before, and my general thought on this is that it is much simpler to use a plugin to get clipboard support: https://github.com/christoomey/vim-system-copy.  That plugin is nice because it supports the system clipboard across all OSes, whereas adding X only helps on Linux.  You may need to install `xclip` if your system doesn't have it, but that is also available as a formula so it's pretty easy to set up.

**Reply by kitmonisit** [reply]
I read through https://github.com/Homebrew/homebrew-core/pull/74263, but I probably missed it, my apologies: Why not have it by default? I disagree with what https://github.com/christoomey/vim-system-copy says about macOS. Getting `macvim` affords me the system clipboard feature IIRC even when using its terminal executable. In Linux, I'd expect the same behavior. What does it break if we have +clipboard by default?

**Reply by danielnachun** [reply]
It doesn't break anything, but adding X as a dependency to a formula drags in many other dependencies, some of which have to be built from source in non-standard install prefixes (which are more common for Homebrew on Linux than macOS).  `vim` is a very popular formula so we'd rather avoid making its installation any more complex than it needs to be.  

System clipboard support for `vim` on the terminal on macOS relies on tools and libraries which are guaranteed to be present (and in a consistent location) on macOS, while system clipboard support for `vim` on the terminal on Linux requires X which we must install through Homebrew.

In the comment below, @carlocab linked to a previous attempt to make a separate formula which builds `vim` with X support.  We are open to the idea of adding this formula but someone has to do the work to set it up.

**Reply by kitmonisit** [reply]
Right, I've been bitten by that host of X dependencies on non-standard install. I'm a non-root Linux user at work.

## Band-aid solution
Hmm, for a start, how about making it a `brew install --with-x vim` option that's documented in its `brew info` entry, complete with a caveat?

## macvim-like solution
In https://github.com/Homebrew/homebrew-core/pull/74263, there is some aversion to the GUI vim "with all the menus and buttons". `macvim` has that approach actually. Calling `macvim` from the shell pulls up GUI vim. With a bit of manual symlinking, I could use the terminal `vim` upon which `macvim` is built. For Linux, that approach is fine with me (though I don't know how much of a bloat GTK is, and whether building `vim-gtk` builds a separate terminal `vim` executable). What do you think?

**Reply by carlocab** [reply]
You can do the band-aid solution by hosting a `vim` formula in your own tap. Options aren't allowed in homebrew/core, unfortunately.

As for the macvim-like solution: it's not so much an aversion to menus and buttons, it's about avoiding dependency bloat. That's why the ideal solution here would be to make it so that you can do `brew install vim`, and then when you want to add a GUI to your existing `vim` install, you do (for example) `brew install vim-gui`.

This is how the `git` formula works. You do `brew install git`, which doesn't install the GUI application. If you want the GUI for Git, you do `brew install git-gui`.

**Reply by danielnachun** [reply]
It is definitely possible to set up the `vim-gtk` formula in such a way that `vim` still launches it in the terminal and then `vim-gtk` or `gvim` launches the GUI version.  As you surmised it's just a matter of setting up symlinks or launch scripts to call `vim` with the appropriate flags.

**Comment by carlocab** [comment]
There was a previous attempt to do something similar: https://github.com/Homebrew/homebrew-core/pull/74263

If you can implement it in the way suggested by [this comment](https://github.com/Homebrew/homebrew-core/pull/74263#issuecomment-811201250), I think that's something we can do in Homebrew.
